### ISSUE TITLE
Crashes on small dataset.

### ISSUE THREAD

**Issue by demikandr** [jcjohnson/torch-rnn#135]
If number of chars is less than 24015, train.lua crashes:

th train.lua -gpu -1 
`Running in CPU mode    
/home/demikandr/torch/install/bin/luajit: bad argument #2 to '?' (end index out of bound at /tmp/luarocks_torch-scm-1-6362/torch7/generic/Tensor.c:969)
stack traceback:
    [C]: at 0x7ff4aea9d3e0
    [C]: in function '__index'
    ./util/DataLoader.lua:35: in function '__init'
    /home/demikandr/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/demikandr/torch/install/share/lua/5.1/torch/init.lua:87>
    [C]: in function 'DataLoader'
    train.lua:76: in main chunk
    [C]: in function 'dofile'
    ...andr/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
    [C]: at 0x00405b80
`

It will be a problem when one do a model sanity check with overfitting.

**Comment by dali99** [comment]
I'm getting a similar issue with even the default dataset
'''
th train.lua \
> -input_h5 data/tiny-shakespeare.h5 \
> -input_json data/tiny-shakespeare.json \
> -gpu -1
Running in CPU mode
/root/torch/install/bin/luajit: bad argument #2 to '?' (end index out of bound at /tmp/luarocks_torch-scm-1-437/torch7/generic/Tensor.c:969)
stack traceback:
        [C]: at 0x7f40929b9b20
        [C]: in function '__index'
        ./util/DataLoader.lua:30: in function '__init'
        /root/torch/install/share/lua/5.1/torch/init.lua:91: in function </root/torch/install/share/lua/5.1/torch/init.lua:87>
        [C]: in function 'DataLoader'
        train.lua:76: in main chunk
        [C]: in function 'dofile'
        /root/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00406670
'''

**Comment by wenjianhugit** [comment]
I also have the same problem...

Running with CUDA on GPU 0	
/home/wenjian/torch/install/bin/luajit: bad argument #2 to '?' (end index out of bound at /tmp/luarocks_torch-scm-1-539/torch7/generic/Tensor.c:992)
stack traceback:
	[C]: at 0x7f75743e44e0
	[C]: in function '__index'
	./util/DataLoader.lua:35: in function '__init'
	/home/wenjian/torch/install/share/lua/5.1/torch/init.lua:91: in function </home/wenjian/torch/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'DataLoader'
	train.lua:76: in main chunk
	[C]: in function 'dofile'
	...jian/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x00406670

Could anyone help ? Thank you !

**Comment by dgcrouse** [comment]
Is anyone still having this issue? Steps to replicate would be great.

**Comment by danieleghisi** [comment]
I've started to have this issue also, out of nothing, and I have no idea why, just using "th train.lua" shows the problem. I'm using torch7 on MacOS 10.10.5. Actual log log is this one:

`MBPdiDaniele2:torch-rnn-master-unicode danieleghisi$ th train.lua -input_h5 data/lowsmall_soundtypes/lowsmall.h5 -input_json data/lowsmall_soundtypes/lowsmall.json -max_epochs 500 -seq_length 40 -num_layers 3 -rnn_size 512 -dropout 0.25 -gpu -1
Running in CPU mode	
/Users/danieleghisi/torch-cl/install/bin/luajit: bad argument #2 to '?' (end index out of bound at /tmp/luarocks_torch-scm-1-4000/torch7/generic/Tensor.c:992)
stack traceback:
	[C]: at 0x05fa1240
	[C]: in function '__index'
	./util/DataLoader.lua:35: in function '__init'
	...nieleghisi/torch-cl/install/share/lua/5.1/torch/init.lua:91: in function <...nieleghisi/torch-cl/install/share/lua/5.1/torch/init.lua:87>
	[C]: in function 'DataLoader'
	train.lua:76: in main chunk
	[C]: in function 'dofile'
	...i/torch-cl/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x0105d81bb0`

**Comment by antihutka** [comment]
DataLoader expects that test and validation sets are both larger than `batch_size * seq_length`. Either increase their sizes when preprocessing or decrease batch size or sequence length when training if you are using small datasets. It should be possible to fix DataLoader to handle this, I might look into it later.
